src/cargo/core/resolver/features.rs

```diff
@@ -0,0 +1,504 @@
+        };
+        r.do_resolve(specs, requested_features)?;
+        log::debug!("features={:#?}", r.activated_features);
+        if r.opts.compare {
+            r.compare();
+        }
+        Ok(ResolvedFeatures {
+            activated_features: r.activated_features,
+            legacy: None,
+            opts: r.opts,
+        })
+    }
+
+    /// Performs the process of resolving all features for the resolve graph.
+    fn do_resolve(
+        &mut self,
+        specs: &[PackageIdSpec],
+        requested_features: &RequestedFeatures,
+    ) -> CargoResult<()> {
+        let member_features = self.ws.members_with_features(specs, requested_features)?;
+        for (member, requested_features) in &member_features {
+            let fvs = self.fvs_from_requested(member.package_id(), requested_features);
+            self.activate_pkg(member.package_id(), &fvs, false)?;
+        }
+        Ok(())
+    }
+
+    fn activate_pkg(
+        &mut self,
+        pkg_id: PackageId,
+        fvs: &[FeatureValue],
+        for_build: bool,
+    ) -> CargoResult<()> {
+        // Add an empty entry to ensure everything is covered. This is intended for
+        // finding bugs where the resolver missed something it should have visited.
+        // Remove this in the future if `activated_features` uses an empty default.
+        self.activated_features
+            .entry((pkg_id, for_build))
+            .or_insert_with(BTreeSet::new);
+        for fv in fvs {
+            self.activate_fv(pkg_id, fv, for_build)?;
+        }
+        if !self.processed_deps.insert((pkg_id, for_build)) {
+            // Already processed dependencies.
+            return Ok(());
+        }
+        for (dep_pkg_id, deps) in self.deps(pkg_id, for_build) {
+            for dep in deps {
+                if dep.is_optional() {
+                    continue;
+                }
+                // Recurse into the dependency.
+                let fvs = self.fvs_from_dependency(dep_pkg_id, dep);
+                self.activate_pkg(
+                    dep_pkg_id,
+                    &fvs,
+                    for_build || (self.opts.decouple_build_deps && dep.is_build()),
+                )?;
+            }
+        }
+        Ok(())
+    }
+
+    /// Activate a single FeatureValue for a package.
+    fn activate_fv(
+        &mut self,
+        pkg_id: PackageId,
+        fv: &FeatureValue,
+        for_build: bool,
+    ) -> CargoResult<()> {
+        match fv {
+            FeatureValue::Feature(f) => {
+                self.activate_rec(pkg_id, *f, for_build)?;
+            }
+            FeatureValue::Crate(dep_name) => {
+                // Activate the feature name on self.
+                self.activate_rec(pkg_id, *dep_name, for_build)?;
+                // Activate the optional dep.
+                for (dep_pkg_id, deps) in self.deps(pkg_id, for_build) {
+                    for dep in deps {
+                        if dep.name_in_toml() == *dep_name {
```

Comment:
I'm personally a fan of reducing rightward-drift wherever possible so if you're also a fan feel free to switch this to a `continue` followed by de-indenting the body